src/OpenTelemetry/Logs/LogRecord.cs

```diff
@@ -213,22 +221,77 @@ public void ForEachScope<TState>(Action<LogRecordScope, TState> callback, TState
             return ref this.Data;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void ResetReferenceCount()
+        {
+            this.PoolReferenceCount = 1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void AddReference()
+        {
+            Interlocked.Increment(ref this.PoolReferenceCount);
```

Comment:
Do we expect multiple threads to update reference count of the same logRecord object simultaneously?

Reset and Increment should always be called from the thread calling `ILogger.Log`.

Decrement which is called by return would also be called from the same thread that called `ILogger.Log` with the only exception being the exporter thread which would also call `Return` but there shouldn't be any other thread operating on the same logRecord object at that time.

Reply:
`AddReference` should always be called serially. But `RemoveReference` there could be racing. Here is what I'm thinking about...

* The thread emitting the log will flow to [OnExport](https://github.com/open-telemetry/opentelemetry-dotnet/blob/2e334df8aec79831ca3660f294dab8fa1e1920d1/src/OpenTelemetry/BatchExportProcessor.cs#L107) which _could_ trigger an export on the dedicated thread. Or the export thread could just decide to wake up right after the log is added to the batch. The emitting thread will call `Return` a bit further along in its call path as will the exporter thread after it exports the log. Those `Return`s could race.

* There could be multiple exporters. Each will fire on its own thread and call `Return` _whenever_ they happen to finish processing. Those `Return`s could race.

So I think using the `Interlocked` operations for reference counting are good to keep it safe?